In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.simplefilter("ignore")

In [ ]:
df=pd.read_csv("train.csv")
pd.set_option('display.max_columns', None)
df.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0              0              0              1              0              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              0              1              0              0              0   
4              0              1              0              0              0   

   ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  ps_ind_14  \
0              0              0              0              0          0   
1              0              0              0              0          0   
2              0              0              0              0          0   
3              0              0              0              0          0   
4              0              0              0              0          0   

   ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_reg_01  \
0         11              0              1              0        0.7   
1          3              0              0              1        0.8   
2         12              1              0              0        0.0   
3          8              1              0              0        0.9   
4          9              1              0              0        0.7   

   ps_reg_02  ps_reg_03  ps_car_01_cat  ps_car_02_cat  ps_car_03_cat  \
0        0.2   0.718070             10              1             -1   
1        0.4   0.766078             11              1             -1   
2        0.0  -1.000000              7              1             -1   
3        0.2   0.580948              7              1              0   
4        0.6   0.840759             11              1             -1   

   ps_car_04_cat  ps_car_05_cat  ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  \
0              0              1              4              1              0   
1              0             -1             11              1              1   
2              0             -1             14              1              1   
3              0              1             11              1              1   
4              0             -1             14              1              1   

   ps_car_09_cat  ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  \
0              0              1             12          2   0.400000   
1              2              1             19          3   0.316228   
2              2              1             60          1   0.316228   
3              3              1            104          1   0.374166   
4              2              1             82          3   0.316070   

   ps_car_13  ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  \
0   0.883679   0.370810   3.605551         0.6         0.5         0.2   
1   0.618817   0.388716   2.449490         0.3         0.1         0.3   
2   0.641586   0.347275   3.316625         0.5         0.7         0.1   
3   0.542949   0.294958   2.000000         0.6         0.9         0.1   
4   0.565832   0.365103   2.000000         0.4         0.6         0.0   

   ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  \
0           3           1          10           1          10           1   
1           2           1           9           5           8           1   
2           2           2           9           1           8           2   
3           2           4           7           1           8           4   
4           2           2           6   

In [ ]:
df.tail()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
595207  1488013       0          3              1         10              0   
595208  1488016       0          5              1          3              0   
595209  1488017       0          1              1         10              0   
595210  1488021       0          5              2          3              1   
595211  1488027       0          0              1          8              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              1              0              0   
595210              0              0              0              1   
595211              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
595207              1              0              0              0   
595208              1              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
595207              0          0         13              1              0   
595208              0          0          6              1              0   
595209              0          0         12              1              0   
595210              0          0         12              1              0   
595211              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
595207              0        0.5        0.3   0.692820             10   
595208              0        0.9        0.7   1.382027              9   
595209              0        0.9        0.2   0.659071              7   
595210              0        0.9        0.4   0.698212             11   
595211              0        0.1        0.2  -1.000000              7   

        ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  ps_car_05_cat  \
595207              1             -1              0              1   
595208              1             -1              0             -1   
595209              1             -1              0             -1   
595210              1             -1              0             -1   
595211              0             -1              0             -1   

        ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
595207              1              1              1              0   
595208             15              0              0              2   
595209              1              1              1              2   
595210             11              1              1              2   
595211              0              1              0              2   

        ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  \
595207              1             31          3   0.374166   0.684631   
595208              1             63          2   0.387298   0.972145   
595209              1             31          3   0.397492   0.596373   
595210              1            101          3   0.374166   0.764434   
595211              1             34          2   0.400000   0.932649   

        ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  \
595207   0.385487   2.645751         0.4         0.5         0.3           3   
595208  -1.000000   3.605551         0.2         0.2         0.0           2   
595209   0.398748   1.732051         0.4         0.0         0.3           3   
595210   0.384968   3.162278         0.0         0.7         0.0           4   
595211   0.378021   3.741657         0.4         0.0         0.5           2   

        ps_calc_05  ps_calc_06  

In [ ]:
df.shape

(595212, 59)

## Observations
*   Features that belong to similar groupings are tagged as such in the feature names (e.g., ind, reg, car, calc).
*   Feature names include the postfix bin to indicate binary features and cat to indicate categorical features.
*   Features without these designations are either continuous or ordinal.
*   There are 595212 rows and 59 columns
*   Target value 1 signifies the claim was filed
*   Each record has a seprate id which does not hold any meaning in prediction







In [ ]:
df.columns

Index(['id', 'target', 'ps_ind_01', 'ps_ind_02_cat', 'ps_ind_03',
       'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_ind_06_bin', 'ps_ind_07_bin',
       'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin', 'ps_ind_11_bin',
       'ps_ind_12_bin', 'ps_ind_13_bin', 'ps_ind_14', 'ps_ind_15',
       'ps_ind_16_bin', 'ps_ind_17_bin', 'ps_ind_18_bin', 'ps_reg_01',
       'ps_reg_02', 'ps_reg_03', 'ps_car_01_cat', 'ps_car_02_cat',
       'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat', 'ps_car_06_cat',
       'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat', 'ps_car_10_cat',
       'ps_car_11_cat', 'ps_car_11', 'ps_car_12', 'ps_car_13', 'ps_car_14',
       'ps_car_15', 'ps_calc_01', 'ps_calc_02', 'ps_calc_03', 'ps_calc_04',
       'ps_calc_05', 'ps_calc_06', 'ps_calc_07', 'ps_calc_08', 'ps_calc_09',
       'ps_calc_10', 'ps_calc_11', 'ps_calc_12', 'ps_calc_13', 'ps_calc_14',
       'ps_calc_15_bin', 'ps_calc_16_bin', 'ps_calc_17_bin', 'ps_calc_18_bin',
       'ps_calc_19_bin', 'ps_calc_20_bin'],


In [ ]:
df.target.value_counts()

0    573518
1     21694
Name: target, dtype: int64

In [ ]:
df.target.value_counts()[1]/df.target.value_counts()[0]

0.037826188541597645

###Data is not balanced and there are significantly more training data for target value 0 than 1

In [ ]:
l=list(df.columns)
c=[] #categorical
b=[] #binary
iv=[] #interval
o=[] #ordinal
for i in l:
  if len(i)>2:
    if i[-3:]=="cat":
      c.append(i)
    elif i[-3:]=="bin":
      b.append(i)
    elif df[i].dtype == float:
      iv.append(i)
    elif df[i].dtype == int:
      o.append(i)
o.remove("target")
len(c),len(b)

(14, 17)

##There are 14 categorical and 17 binary features

# Interval Features

*   There are 11 interval features
*   ind group has no interval features



In [ ]:
df[iv]

ps_reg_01  ps_reg_02  ps_reg_03  ps_car_12  ps_car_13  ps_car_14  \
0             0.7        0.2   0.718070   0.400000   0.883679   0.370810   
1             0.8        0.4   0.766078   0.316228   0.618817   0.388716   
2             0.0        0.0  -1.000000   0.316228   0.641586   0.347275   
3             0.9        0.2   0.580948   0.374166   0.542949   0.294958   
4             0.7        0.6   0.840759   0.316070   0.565832   0.365103   
...           ...        ...        ...        ...        ...        ...   
595207        0.5        0.3   0.692820   0.374166   0.684631   0.385487   
595208        0.9        0.7   1.382027   0.387298   0.972145  -1.000000   
595209        0.9        0.2   0.659071   0.397492   0.596373   0.398748   
595210        0.9        0.4   0.698212   0.374166   0.764434   0.384968   
595211        0.1        0.2  -1.000000   0.400000   0.932649   0.378021   

        ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  
0        3.605551         0.6         0.5         0.2  
1        2.449490         0.3         0.1         0.3  
2        3.316625         0.5         0.7         0.1  
3        2.000000         0.6         0.9         0.1  
4        2.000000         0.4         0.6         0.0  
...           ...         ...         ...         ...  
595207   2.645751         0.4         0.5         0.3  
595208   3.605551         0.2         0.2         0.0  
595209   1.732051         0.4         0.0         0.3  
595210   3.162278         0.0         0.7         0.0  
595211   3.741657         0.4         0.0         0.5  

[595212 rows x 10 columns]

## Ordinal Features

*   There are 15 ordinal features
*   ps_calc_10 have maximum order range of 10



In [ ]:
df[o].describe()

ps_ind_01      ps_ind_03      ps_ind_14      ps_ind_15  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        1.900378       4.423318       0.012451       7.299922   
std         1.983789       2.699902       0.127545       3.546042   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       2.000000       0.000000       5.000000   
50%         1.000000       4.000000       0.000000       7.000000   
75%         3.000000       6.000000       0.000000      10.000000   
max         7.000000      11.000000       4.000000      13.000000   

           ps_car_11     ps_calc_04     ps_calc_05     ps_calc_06  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        2.346072       2.372081       1.885886       7.689445   
std         0.832548       1.117219       1.134927       1.334312   
min        -1.000000       0.000000       0.000000       0.000000   
25%         2.000000       2.000000       1.000000       7.000000   
50%         3.000000       2.000000       2.000000       8.000000   
75%         3.000000       3.000000       3.000000       9.000000   
max         3.000000       5.000000       6.000000      10.000000   

          ps_calc_07     ps_calc_08     ps_calc_09     ps_calc_10  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        3.005823       9.225904       2.339034       8.433590   
std         1.414564       1.459672       1.246949       2.904597   
min         0.000000       2.000000       0.000000       0.000000   
25%         2.000000       8.000000       1.000000       6.000000   
50%         3.000000       9.000000       2.000000       8.000000   
75%         4.000000      10.000000       3.000000      10.000000   
max         9.000000      12.000000       7.000000      25.000000   

          ps_calc_11     ps_calc_12     ps_calc_13     ps_calc_14  
count  595212.000000  595212.000000  595212.000000  595212.000000  
mean        5.441382       1.441918       2.872288       7.539026  
std         2.332871       1.202963       1.694887       2.746652  
min         0.000000       0.000000       0.000000       0.000000  
25%         4.000000       1.000000       2.000000       6.000000  
50%         5.000000       1.000000       3.000000       7.000000  
75%         7.000000       2.000000       4.000000       9.000000  
max        19.000000      10.000000      13.000000      23.000000

## Binary features

*   There are 17 binary features.
*   Target is also binary.



In [ ]:
df[b]

ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0                   0              1              0              0   
1                   0              0              1              0   
2                   0              0              1              0   
3                   1              0              0              0   
4                   1              0              0              0   
...               ...            ...            ...            ...   
595207              0              0              0              1   
595208              0              0              0              1   
595209              1              0              0              0   
595210              0              0              1              0   
595211              1              0              0              0   

        ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  \
0                   0              0              0              0   
1                   0              0              0              0   
2                   0              0              0              0   
3                   0              0              0              0   
4                   0              0              0              0   
...               ...            ...            ...            ...   
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_calc_15_bin  \
0                   0              1              0               0   
1                   0              0              1               0   
2                   1              0              0               0   
3                   1              0              0               0   
4                   1              0              0               0   
...               ...            ...            ...             ...   
595207              1              0              0               0   
595208              1              0              0               1   
595209              1              0              0               0   
595210              1              0              0               0   
595211              1              0              0               0   

        ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
0                    1               1               0               0   
1                    1               1               0               1   
2                    1               1               0               1   
3                    0               0               0               0   
4                    0               0               1               1   
...                ...             ...             ...             ...   
595207               1               1               0               1   
595208               0               1               0               1   
595209               0               1               0               0   
595210               1               1               1               0   
595211               1               0               0               0   

        ps_calc_20_bin  
0                    1  
1                    0  
2                    0  
3                    0  
4                    0  
...                ...  
595207               1  
595208               1  
595209               0  
595210               0  
595211               0  

[595212 rows x 17 columns]

In [ ]:
df.isnull().sum()

id                0
target            0
ps_ind_01         0
ps_ind_02_cat     0
ps_ind_03         0
ps_ind_04_cat     0
ps_ind_05_cat     0
ps_ind_06_bin     0
ps_ind_07_bin     0
ps_ind_08_bin     0
ps_ind_09_bin     0
ps_ind_10_bin     0
ps_ind_11_bin     0
ps_ind_12_bin     0
ps_ind_13_bin     0
ps_ind_14         0
ps_ind_15         0
ps_ind_16_bin     0
ps_ind_17_bin     0
ps_ind_18_bin     0
ps_reg_01         0
ps_reg_02         0
ps_reg_03         0
ps_car_01_cat     0
ps_car_02_cat     0
ps_car_03_cat     0
ps_car_04_cat     0
ps_car_05_cat     0
ps_car_06_cat     0
ps_car_07_cat     0
ps_car_08_cat     0
ps_car_09_cat     0
ps_car_10_cat     0
ps_car_11_cat     0
ps_car_11         0
ps_car_12         0
ps_car_13         0
ps_car_14         0
ps_car_15         0
ps_calc_01        0
ps_calc_02        0
ps_calc_03        0
ps_calc_04        0
ps_calc_05        0
ps_calc_06        0
ps_calc_07        0
ps_calc_08        0
ps_calc_09        0
ps_calc_10        0
ps_calc_11        0


The data has no empty value

In [ ]:
(df==-1).sum().sort_values()

id                     0
ps_car_08_cat          0
ps_car_10_cat          0
ps_car_11_cat          0
ps_car_13              0
ps_car_15              0
ps_calc_01             0
ps_calc_02             0
ps_calc_03             0
ps_calc_04             0
ps_calc_05             0
ps_calc_06             0
ps_calc_07             0
ps_calc_08             0
ps_calc_09             0
ps_calc_10             0
ps_calc_11             0
ps_calc_12             0
ps_calc_13             0
ps_calc_14             0
ps_calc_15_bin         0
ps_calc_16_bin         0
ps_calc_17_bin         0
ps_calc_18_bin         0
ps_calc_19_bin         0
ps_car_06_cat          0
ps_calc_20_bin         0
ps_ind_14              0
ps_ind_06_bin          0
ps_ind_07_bin          0
ps_ind_08_bin          0
ps_ind_09_bin          0
ps_ind_10_bin          0
ps_ind_11_bin          0
ps_ind_12_bin          0
ps_ind_13_bin          0
ps_car_04_cat          0
ps_ind_15              0
ps_ind_17_bin          0
ps_ind_16_bin          0


### Top two features having mising features are:-

*   ps_car_03_cat
*   ps_car_05_cat




##Handling missing values

*   For interval values replace -1 with mean
*   For other values replace -1 with median
*   ps_car_03_cat has median -1 so in this case -1 is repaced by 2



In [ ]:
l=df.columns[(df==-1).sum()>0]
df["ps_car_03_cat"]=df["ps_car_03_cat"].replace(-1,2)
for i in l:
  if i in iv:
    df[i]=df[i].replace(-1,df[i].mean())
  else:
    df[i]=df[i].replace(-1,(df[i].median()))
(df==-1).sum().sort_values()

id                0
ps_car_09_cat     0
ps_car_10_cat     0
ps_car_11_cat     0
ps_car_11         0
ps_car_12         0
ps_car_13         0
ps_car_14         0
ps_car_15         0
ps_calc_01        0
ps_calc_02        0
ps_calc_03        0
ps_calc_04        0
ps_car_08_cat     0
ps_calc_05        0
ps_calc_07        0
ps_calc_08        0
ps_calc_09        0
ps_calc_10        0
ps_calc_11        0
ps_calc_12        0
ps_calc_13        0
ps_calc_14        0
ps_calc_15_bin    0
ps_calc_16_bin    0
ps_calc_17_bin    0
ps_calc_18_bin    0
ps_calc_06        0
ps_calc_19_bin    0
ps_car_07_cat     0
ps_car_05_cat     0
target            0
ps_ind_01         0
ps_ind_02_cat     0
ps_ind_03         0
ps_ind_04_cat     0
ps_ind_05_cat     0
ps_ind_06_bin     0
ps_ind_07_bin     0
ps_ind_08_bin     0
ps_ind_09_bin     0
ps_ind_10_bin     0
ps_ind_11_bin     0
ps_car_06_cat     0
ps_ind_12_bin     0
ps_ind_14         0
ps_ind_15         0
ps_ind_16_bin     0
ps_ind_17_bin     0
ps_ind_18_bin     0


In [ ]:
len(l)

12

Total 12 features have missing values

In [ ]:
df

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0             7       0          2              2          5              1   
1             9       0          1              1          7              0   
2            13       0          5              4          9              1   
3            16       0          0              1          2              0   
4            17       0          0              2          0              1   
...         ...     ...        ...            ...        ...            ...   
595207  1488013       0          3              1         10              0   
595208  1488016       0          5              1          3              0   
595209  1488017       0          1              1         10              0   
595210  1488021       0          5              2          3              1   
595211  1488027       0          0              1          8              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
0                   0              0              1              0   
1                   0              0              0              1   
2                   0              0              0              1   
3                   0              1              0              0   
4                   0              1              0              0   
...               ...            ...            ...            ...   
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              1              0              0   
595210              0              0              0              1   
595211              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
0                   0              0              0              0   
1                   0              0              0              0   
2                   0              0              0              0   
3                   0              0              0              0   
4                   0              0              0              0   
...               ...            ...            ...            ...   
595207              1              0              0              0   
595208              1              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
0                   0          0         11              0              1   
1                   0          0          3              0              0   
2                   0          0         12              1              0   
3                   0          0          8              1              0   
4                   0          0          9              1              0   
...               ...        ...        ...            ...            ...   
595207              0          0         13              1              0   
595208              0          0          6              1              0   
595209              0          0         12              1              0   
595210              0          0         12              1              0   
595211              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
0                   0        0.7        0.2   0.718070             10   
1                   1        0.8        0.4   0.766078             11   
2                   0        0.0        0.0   0.551102              7   
3                   0        0.9        0.2   0.580948              7   
4                   0        0.7        0.6   0.840759             11   
...               .

## Training set is not balanced
class1 is only 3.65%

In [ ]:
class_0, class_1=df.target.value_counts()
(class_1/(class_0+class_1))*100

3.6447517859182947

In [ ]:
new_class1=0.12*(class_0+class_1)
print(new_class1)
print("After achieving a balance of 12% there will be "+ str(new_class1+class_0)+ "records")

71425.44
After achieving a balance of 12% there will be 644943.44records


In [ ]:
new_v=df[df.target==0].sample(class_1)
new_df = pd.concat([df[df.target==1], new_v], axis=0)
new_df.tail()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
275140   687664       0          1              1          6              1   
63370    158676       0          7              1          5              1   
503960  1259398       0          5              2          7              0   
140300   350424       0          0              1          5              0   
502723  1256200       0          0              1          2              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
275140              0              0              1              0   
63370               0              0              0              1   
503960              0              0              0              0   
140300              0              1              0              0   
502723              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
275140              0              0              0              0   
63370               0              0              0              0   
503960              1              0              0              0   
140300              0              0              0              0   
502723              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
275140              0          0         13              1              0   
63370               0          0         13              1              0   
503960              0          0          2              1              0   
140300              0          0          3              0              0   
502723              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
275140              0        0.8        0.6   0.740777              7   
63370               0        0.4        0.0   1.029563             11   
503960              0        0.9        1.6   1.123888             11   
140300              1        0.9        0.2   0.536773              6   
502723              0        0.9        1.3   1.435052             11   

        ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  ps_car_05_cat  \
275140              1              1              1              1   
63370               1              1              1              1   
503960              1              2              0              0   
140300              1              2              0              0   
502723              1              2              0              0   

        ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
275140              1              1              1              0   
63370               3              1              0              2   
503960              0              1              1              2   
140300              0              1              1              0   
502723              6              1              1              2   

        ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  \
275140              1             35          1   0.374166   0.901363   
63370               1            102          2   0.400000   0.966291   
503960              1             98          3   0.400000   0.780893   
140300              1             39          1   0.316070   0.639651   
502723              1             27          3   0.374166   0.723642   

        ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  \
275140   0.304959   3.464102         0.7         0.8         0.0           2   
63370    0.326956   3.605551         0.7         0.1         0.4           2   
503960   0.397492   3.000000         0.9         0.6         0.5           3   
140300   0.374433   2.645751         0.4         0.5         0.8           2   
502723   0.390768   3.162278         0.1         0.8         0.6           3   

        ps_calc_05  ps_calc_06  

## Random under sampling

In [ ]:
class_0, class_1=new_df.target.value_counts()
class_1/(class_0+class_1)

0.5

## Smote Over Sampling

In [ ]:
# import library
from imblearn.over_sampling import SMOTE
from collections import Counter

smote = SMOTE()
x=df.drop(["id","target"], axis=1)
#x=df[["ps_ind_01","ps_ind_02_cat"]]
y=df["target"]
# fit predictor and target variable
x_smote, y_smote = smote.fit_resample(x, y)

print('Original dataset shape', Counter(y))
print('Resample dataset shape', Counter(y_smote))

Original dataset shape Counter({0: 573518, 1: 21694})
Resample dataset shape Counter({0: 573518, 1: 573518})


## No duplicated values

In [ ]:
df.duplicated().any()

False

In [ ]:
df.dtypes

id                  int64
target              int64
ps_ind_01           int64
ps_ind_02_cat       int64
ps_ind_03           int64
ps_ind_04_cat       int64
ps_ind_05_cat       int64
ps_ind_06_bin       int64
ps_ind_07_bin       int64
ps_ind_08_bin       int64
ps_ind_09_bin       int64
ps_ind_10_bin       int64
ps_ind_11_bin       int64
ps_ind_12_bin       int64
ps_ind_13_bin       int64
ps_ind_14           int64
ps_ind_15           int64
ps_ind_16_bin       int64
ps_ind_17_bin       int64
ps_ind_18_bin       int64
ps_reg_01         float64
ps_reg_02         float64
ps_reg_03         float64
ps_car_01_cat       int64
ps_car_02_cat       int64
ps_car_03_cat       int64
ps_car_04_cat       int64
ps_car_05_cat       int64
ps_car_06_cat       int64
ps_car_07_cat       int64
ps_car_08_cat       int64
ps_car_09_cat       int64
ps_car_10_cat       int64
ps_car_11_cat       int64
ps_car_11           int64
ps_car_12         float64
ps_car_13         float64
ps_car_14         float64
ps_car_15   

In [ ]:
df[iv].corr()

ps_reg_01  ps_reg_02  ps_reg_03  ps_car_12  ps_car_13  ps_car_14  \
ps_reg_01    1.000000   0.471027   0.372227   0.018782   0.027097   0.004791   
ps_reg_02    0.471027   1.000000   0.749089   0.171453   0.194316   0.045006   
ps_reg_03    0.372227   0.749089   1.000000   0.189279   0.223006   0.055556   
ps_car_12    0.018782   0.171453   0.189279   1.000000   0.672014   0.470525   
ps_car_13    0.027097   0.194316   0.223006   0.672014   1.000000   0.330308   
ps_car_14    0.004791   0.045006   0.055556   0.470525   0.330308   1.000000   
ps_car_15    0.002765   0.054510   0.071936   0.049739   0.529519  -0.009357   
ps_calc_01  -0.000891   0.001347  -0.000058  -0.000721   0.000157  -0.003136   
ps_calc_02   0.002194  -0.000770   0.000558  -0.000392   0.000040  -0.000845   
ps_calc_03  -0.001538   0.000022   0.001080   0.000151  -0.000200  -0.000642   

            ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  
ps_reg_01    0.002765   -0.000891    0.002194   -0.001538  
ps_reg_02    0.054510    0.001347   -0.000770    0.000022  
ps_reg_03    0.071936   -0.000058    0.000558    0.001080  
ps_car_12    0.049739   -0.000721   -0.000392    0.000151  
ps_car_13    0.529519    0.000157    0.000040   -0.000200  
ps_car_14   -0.009357   -0.003136   -0.000845   -0.000642  
ps_car_15    1.000000    0.000344   -0.000252   -0.001043  
ps_calc_01   0.000344    1.000000   -0.002418    0.000526  
ps_calc_02  -0.000252   -0.002418    1.000000    0.002948  
ps_calc_03  -0.001043    0.000526    0.002948    1.000000

In [ ]:
df[o].corr()

ps_ind_01  ps_ind_03  ps_ind_14  ps_ind_15  ps_car_11  \
ps_ind_01    1.000000   0.223408  -0.024971   0.122137   0.008262   
ps_ind_03    0.223408   1.000000   0.001990   0.170449  -0.059335   
ps_ind_14   -0.024971   0.001990   1.000000  -0.121547   0.004415   
ps_ind_15    0.122137   0.170449  -0.121547   1.000000   0.023041   
ps_car_11    0.008262  -0.059335   0.004415   0.023041   1.000000   
ps_calc_04  -0.000943   0.000703  -0.001102   0.001584   0.000240   
ps_calc_05  -0.001171  -0.001245   0.000205   0.000874   0.002456   
ps_calc_06   0.000147   0.002361  -0.000242   0.000033  -0.000242   
ps_calc_07  -0.000658   0.001533   0.001023   0.001721  -0.000330   
ps_calc_08   0.002993   0.001267   0.004077  -0.001309  -0.002432   
ps_calc_09  -0.000085  -0.001544  -0.001443  -0.000985  -0.000476   
ps_calc_10  -0.001345   0.000230   0.001976  -0.001279  -0.000582   
ps_calc_11  -0.001263  -0.002716  -0.000243  -0.001952  -0.001190   
ps_calc_12   0.001521  -0.002352   0.000284  -0.001263  -0.001203   
ps_calc_13   0.000199   0.000133   0.000307  -0.000922   0.001160   
ps_calc_14  -0.000891   0.000205  -0.000387   0.001348   0.000931   

              ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  \
ps_ind_01  -9.430788e-04   -0.001171    0.000147   -0.000658    0.002993   
ps_ind_03   7.029223e-04   -0.001245    0.002361    0.001533    0.001267   
ps_ind_14  -1.102298e-03    0.000205   -0.000242    0.001023    0.004077   
ps_ind_15   1.583613e-03    0.000874    0.000033    0.001721   -0.001309   
ps_car_11   2.404439e-04    0.002456   -0.000242   -0.000330   -0.002432   
ps_calc_04  1.000000e+00    0.000633   -0.001845    0.001445   -0.001847   
ps_calc_05  6.326671e-04    1.000000   -0.001394   -0.002212   -0.001792   
ps_calc_06 -1.845321e-03   -0.001394    1.000000   -0.000075   -0.000603   
ps_calc_07  1.445124e-03   -0.002212   -0.000075    1.000000   -0.002085   
ps_calc_08 -1.846555e-03   -0.001792   -0.000603   -0.002085    1.000000   
ps_calc_09 -3.835483e-05    0.001154    0.001710    0.000756    0.000013   
ps_calc_10  5.835325e-08   -0.000236   -0.001534   -0.000530   -0.002666   
ps_calc_11  2.445677e-04    0.000387   -0.000575   -0.000109    0.002016   
ps_calc_12  9.784493e-04   -0.000212    0.000607   -0.000389    0.000109   
ps_calc_13 -2.146282e-04    0.000270    0.000974   -0.000599   -0.000304   
ps_calc_14  1.179561e-03   -0.000969   -0.000870    0.001908   -0.000171   

            ps_calc_09    ps_calc_10  ps_calc_11  ps_calc_12  ps_calc_13  \
ps_ind_01    -0.000085 -1.344982e-03   -0.001263    0.001521    0.000199   
ps_ind_03    -0.001544  2.302873e-04   -0.002716   -0.002352    0.000133   
ps_ind_14    -0.001443  1.975748e-03   -0.000243    0.000284    0.000307   
ps_ind_15    -0.000985 -1.278706e-03   -0.001952   -0.001263   -0.000922   
ps_car_11    -0.000476 -5.816562e-04   -0.001190   -0.001203    0.001160   
ps_calc_04   -0.000038  5.835325e-08    0.000245    0.000978   -0.000215   
ps_calc_05    0.001154 -2.358151e-04    0.000387   -0.000212    0.000270   
ps_calc_06    0.001710 -1.534410e-03   -0.000575    0.000607    0.000974   
ps_calc_07    0.000756 -5.302752e-04   -0.000109   -0.000389   -0.000599   
ps_calc_08    0.000013 -2.665669e-03    0.002016    0.000109   -0.000304   
ps_calc_09    1.000000 -1.819156e-04   -0.000898   -0.001409   -0.000630   
ps_calc_10   -0.000182  1.000000e+00    0.000394    0.002086    0.001279   
ps_calc_11   -0.000898  3.942191e-04    1.000000    0.001036    0.003702   
ps_calc_12   -0.001409  2.086146e-03    0.001036    1.000000   -0.000300   
ps_calc_13   -0.000630  1.279018e-03    0.003702   -0.000300    1.000000   
ps_calc_14   -0.000066  1.252413e-03    0.000480    0.000980   -0.002325   

            ps_calc_14  
ps_ind_01    -0.000891  
ps_ind_03     0.000205  
ps_ind_14    -0.000387  
ps_ind_15     0.001348  
ps_car_11     0.000931  
ps_calc_04    0.001180  
ps_calc_05   -0.000969  
ps_calc_06   -0.000870  
ps_calc_07    0.001908  
ps

## One Hot Encoding

In [ ]:
from sklearn.preprocessing import OneHotEncoder
for i in c:
  en=OneHotEncoder()
  f=np.array(new_df[i]).reshape(-1,1)
  en_df=en.fit_transform(f).toarray()
  col=[i+'_'+j for j in list(en.categories_[0].astype(str))]
  en_df=pd.DataFrame(en_df, columns=col, index=new_df.index)
  
  en_df=new_df.join(en_df)
en_df=en_df.drop(c,axis=1)
en_df.shape

(43388, 149)

In [ ]:
en_df.shape[1]-new_df.shape[1]

90

90 features are increased after one hot encoding

##Standard Scaler

In [ ]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
#interval and categorical

## Summary 

*   Data is imbalaced
*   Features that belong to similar groupings are tagged as such in the feature names (e.g., ind, reg, car, calc).
*   Feature names include the postfix bin to indicate binary features and cat to indicate categorical features.
*   Features without these designations are either continuous or ordinal.
*   There are 595212 rows and 59 columns
*   Target value 1 signifies the claim was filed
*   Each record has a seprate id which does not hold any meaning in prediction
*   Missing values are in the form of -1
*   There are 14 categorical, 17 binary, 17 ordinal and 11 interval values






##Train test Split

In [ ]:
from sklearn.model_selection import train_test_split
x=df.drop(["id","target"], axis=1)
y=df["target"]
X_train, X_test, y_train, y_test=train_test_split(x,y, test_size=0.2)

In [ ]:
X_train.shape

(476169, 57)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

In [ ]:
m=LogisticRegression(max_iter=10000)
m.fit(X_train, y_train)

LogisticRegression(max_iter=10000)

##F1 score is 0 which suggest it is a bad classifier.
###The possible reason for this score is the huge imbalance between target values

In [ ]:
from sklearn.metrics import f1_score
y_pred=m.predict(X_train)
f1_score(y_train,y_pred)

0.0

In [ ]:
accuracy_score(y_train, y_pred)

0.9637502651369577

##Now we use only 1 feature that should significantly decrease accuracy score.

In [ ]:
x=df["ps_ind_01"]
m=LogisticRegression(max_iter=10000)
m.fit(X_train, y_train)
y_pred_one_feat=m.predict(X_train)
accuracy_score(y_train, y_pred_one_feat)

0.9637502651369577

Accuracy score remains same so it is misleading and should not be use as a metric to measure performance.

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98    458908
           1       0.00      0.00      0.00     17261

    accuracy                           0.96    476169
   macro avg       0.48      0.50      0.49    476169
weighted avg       0.93      0.96      0.95    476169



Due to less training data for target value 1 precision, recall and f1 score is 0

In [ ]:
m=LinearSVC()
X_train=X_train[:100000]
y_train=y_train[:100000]
m.fit(X_train, y_train)
y_pred=m.predict(X_train)
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98     96351
           1       0.00      0.00      0.00      3649

    accuracy                           0.96    100000
   macro avg       0.48      0.50      0.49    100000
weighted avg       0.93      0.96      0.95    100000



In [ ]:
import xgboost as xgb
model =xgb.XGBClassifier( 
                      learning_rate=0.06,  
                      colsample_bytree = 0.6,
                      subsample = 1, 
                      n_estimators=600,                       
                      max_depth=3, 
                      gamma=0)
model.fit(X_train, y_train)
y_pred=model.predict(X_train)
print(classification_report(y_train, y_pred))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98    458832
           1       1.00      0.00      0.00     17337

    accuracy                           0.96    476169
   macro avg       0.98      0.50      0.49    476169
weighted avg       0.96      0.96      0.95    476169



If not marking negative sample as positive is top priority then all three works same  with precision of 0.96

If not missing a positive sample is the priority then Xgboost is best so far with precision of 1

### Using under samplind data i.e. new_df

In [ ]:
from sklearn.metrics import f1_score
x=new_df.drop(["id","target"], axis=1)
y=new_df["target"]
X_train_tr, X_test_tr, y_train_tr, y_test_tr=train_test_split(x,y, test_size=0.2)
m=LogisticRegression(max_iter=10000)
m.fit(X_train_tr, y_train_tr)
y_pred=m.predict(X_train_tr)
f1_score(y_train_tr,y_pred)

0.5695720502901354

F1 score is considerably increased when using balance data.

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_train_tr, y_pred))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00     17358
           1       0.50      1.00      0.67     17352

    accuracy                           0.50     34710
   macro avg       0.25      0.50      0.33     34710
weighted avg       0.25      0.50      0.33     34710



In [ ]:
m=LinearSVC()
m.fit(X_train_tr, y_train_tr)
y_pred=m.predict(X_train_tr)
f1_score(y_train_tr,y_pred)

0.6665898351964965

In [ ]:
from sklearn.model_selection import GridSearchCV
p = { 
    'max_iter': [50, 200, 400, 600, 800, 1000],
    'penalty' : ["l1","l2"],
     "loss" : ["hinge", "Squared_hinge"],
}
g_cv=GridSearchCV(estimator=LinearSVC(), param_grid=p, cv=5)
g_cv.fit(X_train_tr, y_train_tr)
g_crf= g_cv.best_estimator_
g_cv.best_params_

{'loss': 'hinge', 'max_iter': 800, 'penalty': 'l2'}

In [ ]:
y_pred=g_crf.predict(X_train_tr)
f1_score(y_train_tr,y_pred)

0.6415147462518477

In [ ]:
y_pred=g_crf.predict(X_test_tr)
f1_score(y_test_tr,y_pred)

0.6418354077651867

In [ ]:
import xgboost as xgb
model =xgb.XGBClassifier( 
                      learning_rate=0.06,  
                      colsample_bytree = 0.6,
                      subsample = 0.8, 
                      n_estimators=200,                       
                      max_depth=3, 
                      gamma=0)
model.fit(X_train_tr, y_train_tr)
y_pred=model.predict(X_train_tr)
print(classification_report(y_train_tr, y_pred))

              precision    recall  f1-score   support

           0       0.61      0.64      0.62     17386
           1       0.62      0.59      0.60     17324

    accuracy                           0.61     34710
   macro avg       0.62      0.61      0.61     34710
weighted avg       0.62      0.61      0.61     34710



In [ ]:
f1_score(y_train_tr,y_pred)

0.6045506997662515

xgb classifier has less f1 score due to limited data as majority of data was removed in over sampling

##Using encoded data i.e. en_df

In [ ]:
en_df.head()

id  target  ps_ind_01  ps_ind_03  ps_ind_06_bin  ps_ind_07_bin  \
9    28       1          1          2              0              1   
19   64       1          0          2              1              0   
28   84       1          0          0              1              0   
39  107       1          2          3              0              0   
41  110       1          5          1              0              0   

    ps_ind_08_bin  ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
9               0              0              0              0              0   
19              0              0              0              0              0   
28              0              0              0              0              0   
39              1              0              0              0              0   
41              1              0              0              0              0   

    ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
9               0          0          4              0              0   
19              0          0          8              1              0   
28              0          0          8              1              0   
39              0          0         12              1              0   
41              0          0          8              1              0   

    ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_11  ps_car_12  \
9               1        0.9        1.4   2.316652          2   0.447214   
19              0        0.9        0.3   0.402337          3   0.315278   
28              0        0.2        0.3   0.551102          2   0.447214   
39              0        0.3        0.0   0.920937          3   0.316228   
41              0        0.8        0.3   0.812404          2   0.387298   

    ps_car_13  ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  \
9    1.458184   0.390256   3.605551         0.7         0.8         0.8   
19   0.464593   0.361248   0.000000         0.7         0.3         0.2   
28   0.720265   0.441588   2.645751         0.8         0.4         0.1   
39   0.740165   0.359861   3.464102         0.7         0.3         0.6   
41   1.023898   0.408656   3.605551         0.1         0.7         0.5   

    ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  \
9            2           2           8           2           9           1   
19           2           2           8           2           5           2   
28           2           2          10           3           9           2   
39           4           0           9           4          10           2   
41           4           1           8           1          10           4   

    ps_calc_10  ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  \
9           11           3           5           0           6   
19          10          10           3           1          11   
28           5           3           2           0           8   
39          10           6           0           8          14   
41           6           5           2           0           7   

    ps_calc_15_bin  ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  \
9                0               1               0               0   
19               0               1               1               0   
28               0               1               1               0   
39               0               1               0               1   
41               1               1               0               0   

    ps_calc_19_bin  ps_calc_20_bin  ps_car_11_cat_1  ps_car_11_cat_2  \
9                1               0              0.0              0.0   
19               1               0              0.0              0.0   
28               0               0              0.0              0.0   
39               0               0              0.0              0.0   
41               0               0              0.0              0.0   

    ps_car_11_cat_3  ps_

In [ ]:
x=en_df.drop(["id","target"], axis=1)
y=en_df["target"]
X_train_en, X_test_en, y_train_en, y_test_en=train_test_split(x,y, test_size=0.2)

In [ ]:
m=LogisticRegression(max_iter=10000)
m.fit(X_train_en, y_train_en)
y_pred=m.predict(X_train_en)
f1_score(y_train_en,y_pred)

0.5607853765772277

In [ ]:
m=LinearSVC(loss= 'hinge', max_iter= 600, penalty= 'l2')
m.fit(X_train_en, y_train_en)
y_pred=m.predict(X_train_en)
f1_score(y_train_en,y_pred)

0.55929288967993

Using encoded data increases the performance of the model.

In [ ]:
model =xgb.XGBClassifier( 
                      learning_rate=0.06,  
                      colsample_bytree = 0.6,
                      subsample = 0.8, 
                      n_estimators=200,                       
                      max_depth=3, 
                      gamma=0)
model.fit(X_train_en, y_train_en)
y_pred=model.predict(X_train_en)
f1_score(y_train_en,y_pred)

0.5977769852591536

In [ ]:
import xgboost as xgb
from sklearn.model_selection import GridSearchCV
p = { 
    'learning_rate': [0.01, 0.05, 0.1, 0.4, 0.6, 1],
    'n_estimators' : [200, 400, 600, 800, 1000],
     "gamma" : [0,0.5,1],
     "max_depth": [3,5,7],
     "subsample": [0.4, 0.6, 0.8, 1]
}
g_xg=GridSearchCV(estimator=xgb.XGBClassifier(), param_grid=p,n_jobs=-1, cv=5)
X_train_sh=X_train_tr[:500]
y_train_sh=y_train_tr[:500]
g_xg.fit(X_train_sh, y_train_sh)
g_bxg= g_xg.best_estimator_
g_xg.best_params_

{'gamma': 0,
 'learning_rate': 0.01,
 'max_depth': 3,
 'n_estimators': 600,
 'subsample': 1}

In [ ]:
y_pred=g_bxg.predict(X_train_tr)
f1_score(y_train_tr,y_pred)

0.5507169566881606

In [ ]:
y_pred=g_bxg.predict(X_test_tr)
f1_score(y_test_tr,y_pred)

0.5437981779957953

Xgboost will have more performance if more data is used for training.

In [ ]:
from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier
clf = RandomForestClassifier()
ada_m=AdaBoostClassifier(base_estimator=clf)
ada_m.fit(X_train_en, y_train_en)
y_pred=ada_m.predict(X_train_en)
f1_score(y_train_en,y_pred)

1.0

In [ ]:
print(classification_report(y_train_en, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     17404
           1       1.00      1.00      1.00     17306

    accuracy                           1.00     34710
   macro avg       1.00      1.00      1.00     34710
weighted avg       1.00      1.00      1.00     34710



In [ ]:
y_pred=ada_m.predict(X_test_en)
f1_score(y_test_en,y_pred)

0.5618823529411764

Ada boost overfits the training data

## Mlp Classifier
We can choose the right number of layers and size of hidden layer by hit and trail method i.e. by choosing different number of hidden layer and number of neurons till we find the best one.

In [ ]:
from sklearn.neural_network import MLPClassifier
mlp_m=MLPClassifier()
mlp_m.fit(X_train_en, y_train_en)
y_pred=mlp_m.predict(X_train_en)
f1_score(y_train_en,y_pred)

0.7347564792788281

In [ ]:
y_pred=mlp_m.predict(X_test_en)
f1_score(y_test_en,y_pred)

0.5090819833087874

In [ ]:
from sklearn.neural_network import MLPClassifier
mlp_m=MLPClassifier(hidden_layer_sizes=(150,100,50))
mlp_m.fit(X_train_en, y_train_en)
y_pred=mlp_m.predict(X_train_en)
print("f1 score on training", f1_score(y_train_en,y_pred))
y_pred=mlp_m.predict(X_test_en)
f1_score(y_test_en,y_pred)

f1 score on training 0.9803662667202481


0.5350073521094899

In [ ]:
from sklearn.neural_network import MLPClassifier
mlp_m=MLPClassifier(hidden_layer_sizes=(50,50,50))
mlp_m.fit(X_train_en, y_train_en)
y_pred=mlp_m.predict(X_train_en)
print("f1 score on training", f1_score(y_train_en,y_pred))
y_pred=mlp_m.predict(X_test_en)
f1_score(y_test_en,y_pred)

f1 score on training 0.801413766249326


0.5139304077484156

In [ ]:
from sklearn.neural_network import MLPClassifier
mlp_m=MLPClassifier(hidden_layer_sizes=(150,150,100,50,50))
mlp_m.fit(X_train_en, y_train_en)
y_pred=mlp_m.predict(X_train_en)
print("f1 score on training", f1_score(y_train_en,y_pred))
y_pred=mlp_m.predict(X_test_en)
f1_score(y_test_en,y_pred)

f1 score on training 0.9831454210768522


0.541389045736872

In [ ]:
from sklearn.neural_network import MLPClassifier
mlp_m=MLPClassifier(hidden_layer_sizes=(100,50))
mlp_m.fit(X_train_en, y_train_en)
y_pred=mlp_m.predict(X_train_en)
print("f1 score on training", f1_score(y_train_en,y_pred))
y_pred=mlp_m.predict(X_test_en)
f1_score(y_test_en,y_pred)

f1 score on training 0.8629855667717582


0.5285436448918888

Mlp classifier with 4 layers works best

## Using Smote data over sampled

In [ ]:
X_train_sm, X_test_sm, y_train_sm, y_test_sm=train_test_split(x_smote, y_smote, test_size=0.2)

In [ ]:
m=LogisticRegression(max_iter=10000)
m.fit(X_train_sm, y_train_sm)
y_pred=m.predict(X_train_sm)
f1_score(y_train_sm,y_pred)

0.8939600123260464

In [ ]:
y_pred=m.predict(X_test_sm)
f1_score(y_test_sm,y_pred)

0.893582760218241

In [ ]:
m_svc=LinearSVC(loss= 'hinge', max_iter= 600, penalty= 'l2')
m_svc.fit(X_train_sm, y_train_sm)
y_pred=m_svc.predict(X_train_sm)
f1_score(y_train_sm,y_pred)

0.8304014865912012

In [ ]:
y_pred=m_svc.predict(X_test_sm)
f1_score(y_test_sm,y_pred)

0.8302397428945261

In [ ]:
model =xgb.XGBClassifier( 
                      learning_rate=0.01,  
                      colsample_bytree = 0.6,
                      subsample = 1, 
                      n_estimators=600,                       
                      max_depth=3, 
                      gamma=0)
model.fit(X_train_sm, y_train_sm)
y_pred=model.predict(X_train_sm)
f1_score(y_train_sm,y_pred)

0.9432276826464393

In [ ]:
y_pred=model.predict(X_test_sm)
f1_score(y_test_sm,y_pred)

0.9428131343708822

Xgboost performs best